### PR TITLE
Distributed image import

### PR DESCRIPTION
**Pull request by fmount**
This patch introduces the required changes to support `distributed_image_import`.
Distributed image import `requires > 1` `glanceAPI` replicas and it is possible to redirect the import operation
to a given worker via `worker_self_reference_url`.
To properly set that value, however, we need to resolve the `POD` `IP` addresses.
`StatefulSet` provides stable network access to the Pod/container/replicas, and by adding a `headless service`
beside the existing internal/external `Services` we can resolve the Pod IP via `DNS`  even though it's scheduled on
a different node.
The additional `Service` is only used for pod2pod communication.

Implements: [OSPRH-1243](https://issues.redhat.com/browse/OSPRH-1243)

### CODE REVIEW

**Review by konan-abhi — COMMENTED** [review]
Looks good, few nits!!
Great work!!

File: templates/common/bin/kolla_extend_start [hunk 1]
```diff
@@ -0,0 +1,24 @@
+#!/bin/bash
+
+GLANCE_PORT=${GLANCE_PORT:-9292}
+GLOBAL_CONF=${GLOBAL_CONF:-/etc/glance/glance.conf.d/01-config.conf}
+
+function set_worker_self_url {
+cat <<EOF > "${GLOBAL_CONF}"
+[DEFAULT]
+worker_self_reference_url=$(hostname).${GLANCE_DOMAIN}:${GLANCE_PORT}
```
Comment (fultonj) [inline templates/common/bin/kolla_extend_start:9]: need to insert `http://` here

Comment (fmount) [reply templates/common/bin/kolla_extend_start]: Done

File: templates/common/bin/kolla_extend_start [hunk 2]
```diff
@@ -0,0 +1,24 @@
+#!/bin/bash
+
+GLANCE_PORT=${GLANCE_PORT:-9292}
+GLOBAL_CONF=${GLOBAL_CONF:-/etc/glance/glance.conf.d/01-config.conf}
+
+function set_worker_self_url {
```
[the remark this answers is not in the document]
Comment (fmount) [reply templates/common/bin/kolla_extend_start]: Done

File: controllers/glanceapi_controller.go [hunk 3]
```diff
@@ -440,6 +439,9 @@ func (r *GlanceAPIReconciler) reconcileInit(
 				})
 			}
 		}
+		// register the service hostname as base domain to build the work_self_url
+		// that will be used for distributed image import across multiple replicas
```
[the remark this answers is not in the document]
Comment (fmount) [reply controllers/glanceapi_controller.go:443]: Done

File: api/v1beta1/glanceapi_types.go [hunk 4]
```diff
@@ -95,6 +95,12 @@ type GlanceAPIStatus struct {
 
 	// NetworkAttachments status of the deployment pods
 	NetworkAttachments map[string][]string `json:"networkAttachments,omitempty"`
+
+	// Domain is a parameter used by each glanceAPI replicas to setup a worker
+	// and set the worker_self_url config option. It's requierd when distributed
```
[the remark this answers is not in the document]
Comment (fmount) [reply api/v1beta1/glanceapi_types.go]: Done

**Comment by fmount** [comment]
Envtests cover the headless service creation, in a follow up we are going to add documentation, scripts to test this feature, and a potential kuttl test.

**Comment by fmount** [comment]
From the CI execution we can see the following data:
1.  For each `StatefulSet` an associated `headless Service` is [created and properly annotated](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openstack-k8s-operators_glance-operator/443/pull-ci-openstack-k8s-operators-glance-operator-main-glance-operator-build-deploy-tempest/1758436406973173760/artifacts/glance-operator-build-deploy-tempest/openstack-k8s-operators-gather/artifacts/must-gather/quay-io-openstack-k8s-operators-openstack-must-gather-sha256-f745587ef185f9023b8c83c4d40a4e1c7478de462a91dc6280e5277545650bbe/namespaces/openstack/services/glance-default-internal-api.yaml)

2. [container-scripts/kolla_extend_start](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openstack-k8s-operators_glance-operator/443/pull-ci-openstack-k8s-operators-glance-operator-main-glance-operator-build-deploy-tempest/1758436406973173760/artifacts/glance-operator-build-deploy-tempest/openstack-k8s-operators-gather/artifacts/must-gather/quay-io-openstack-k8s-operators-openstack-must-gather-sha256-f745587ef185f9023b8c83c4d40a4e1c7478de462a91dc6280e5277545650bbe/namespaces/openstack/secrets/glance/glance-scripts.yaml) is properly mounted to the `glance-api` container

3.  [GLANCE_DOMAIN](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openstack-k8s-operators_glance-operator/443/pull-ci-openstack-k8s-operators-glance-operator-main-glance-operator-build-deploy-tempest/1758436406973173760/artifacts/glance-operator-build-deploy-tempest/openstack-k8s-operators-gather/artifacts/must-gather/quay-io-openstack-k8s-operators-openstack-must-gather-sha256-f745587ef185f9023b8c83c4d40a4e1c7478de462a91dc6280e5277545650bbe/namespaces/openstack/pods/glance-default-external-api-0/glance-default-external-api-0-describe) is provided as an env variable, and used to set `worker_self_reference_url` within each replica

4. from the [glance config secret](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openstack-k8s-operators_glance-operator/443/pull-ci-openstack-k8s-operators-glance-operator-main-glance-operator-build-deploy-tempest/1758436406973173760/artifacts/glance-operator-build-deploy-tempest/openstack-k8s-operators-gather/artifacts/must-gather/quay-io-openstack-k8s-operators-openstack-must-gather-sha256-f745587ef185f9023b8c83c4d40a4e1c7478de462a91dc6280e5277545650bbe/namespaces/openstack/secrets/glance/glance-default-external-config-data.yaml), `glance-direct` is enabled by default, and it allow us to take advantage of this feature

We'll follow up with a doc patch and any potential kuttl test if we want to do more, but we should be good here.

**Review by fultonj — COMMENTED** [review]
+2 thank you @fmount

**Review by konan-abhi — APPROVED** [review]
Thank you!

File: controllers/glance_common.go [hunk 5]
```diff
@@ -138,3 +142,79 @@ func GenerateConfigsGeneric(
...
+	instance *glancev1.GlanceAPI,
+	serviceLabels map[string]string,
+) (ctrl.Result, error) {
+
+	endpointName := instance.Name
+	// The endpointName for headless services **must** match with:
+	// - statefulset.metadata.name
+	// - statefulset.spec.servicename
+	if instance.Spec.APIType != "single" {
+		endpointName = instance.Name + "-api"
```
[the remark this answers is not in the document]
Comment (fmount) [reply controllers/glance_common.go:159]: right, even "single" is present as `const` already: I'll fix that as part of the follow up doc change.

**Pull request merged**